src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -0,0 +1,593 @@
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.NotThreadSafe;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
+
+/**
+ * Scanner that will scan down all forks when we hit parallel blocks before continuing, but generally runs in linear order
+ * <p>Think of it as the opposite of {@link DepthFirstScanner}.
+ *
+ * <p>This is a fairly efficient way to visit all FlowNodes, and provides three useful guarantees:
+ * <ul>
+ *   <li>Every FlowNode is visited, and visited EXACTLY ONCE (not true for LinearScanner)</li>
+ *   <li>All parallel branches are visited before we move past the parallel block (not true for DepthFirstScanner)</li>
+ *   <li>For EVERY block, the BlockEndNode is visited before the BlockStartNode (not true for DepthFirstScanner, with parallels)</li>
+ * </ul>
+ *
+ * <p>The big advantages of this approach:
+ * <ul>
+ *     <li>Blocks are visited in the order they end (no backtracking) - helps with working a block at a time</li>
+ *     <li>Points are visited in linear order within a block (easy to use for analysis)</li>
+ *     <li>Minimal state information needed</li>
+ *     <li>Branch information is available for use here</li>
+ * </ul>
+ *
+ * @author Sam Van Oort
+ */
+@NotThreadSafe
+public class ForkScanner extends AbstractFlowScanner {
+
+    @CheckForNull
+    public NodeType getCurrentType() {
+        return currentType;
+    }
+
+    @CheckForNull
+    public NodeType getNextType() {
+        return nextType;
+    }
+
+    /** Used to recognize special nodes */
+    public enum NodeType {
+        /** Not any of the parallel types */
+        NORMAL,
+        /**{@link BlockStartNode} starting a parallel block */
+        PARALLEL_START,
+        /**{@link BlockEndNode} ending a parallel block */
+        PARALLEL_END,
+        /**{@link BlockStartNode} starting a branch of a parallel */
+        PARALLEL_BRANCH_START,
+        /**{@link BlockEndNode} ending a parallel block... or last executed nodes */
+        PARALLEL_BRANCH_END,
+    }
+
+    // Last element in stack is end of myCurrent parallel start, first is myCurrent start
+    ArrayDeque<ParallelBlockStart> parallelBlockStartStack = new ArrayDeque<ParallelBlockStart>();
+
+    /** FlowNode that will terminate the myCurrent parallel block */
+    FlowNode currentParallelStartNode = null;
+
+    ParallelBlockStart currentParallelStart = null;
+
+    private boolean walkingFromFinish = false;
+
+    NodeType currentType = null;
+    NodeType nextType = null;
+
+    public ForkScanner() {
+
```

Comment:
@oleg-nenashev I believe it was (check back at the original) -- you can invoke methods on a ForkScanner with no nodes feeding it, they'll return null, empty, or an exception as appropriate.